File: apps/desktop/src/features/project/sessionDocument.ts
import type { ChannelKind, ChannelRole, ChannelState, FxUnitState, MixerSnapshot, RecordedTakeState, RecordingTap } from "../../adapters/MixerControlPort";

export interface ProjectSessionDocument {
  schemaVersion: 1;
  projectId: string;
  media: Array<{ id: string; path: string; missing: boolean }>;
  channels: ProjectSessionChannel[];
  fxUnits: ProjectSessionFxUnit[];
  fxSends: ProjectSessionFxSend[];
  channelHarmony: ProjectSessionHarmony[];
  plugins: unknown[];
  recordedTakes: ProjectSessionRecordedTake[];
}

export interface ProjectSessionChannel {
  id: string;
  name: string;
  kind: string;
  role: string;
  sourceUid: string;
  enabled: boolean;
  muted: boolean;
  solo: boolean;
  monitor: boolean;
  recordArm: boolean;
  eqEnabled: boolean;
  noiseEnabled: boolean;
  compEnabled: boolean;
  deEsserEnabled: boolean;
  insertFxEnabled: boolean;
  gainDb: number;
  faderDb: number;
  pan: number;
  noiseThresholdDb: number;
  noiseRangeDb: number;
  noiseHoldMs: number;
  noiseReleaseMs: number;
  compThresholdDb: number;
  compRatio: number;
  compAttackMs: number;
  compReleaseMs: number;
}

export interface ProjectSessionFxUnit {
  unitId: string;
  programId: number;
  processorType: string;
  revision: number;
  enabled: boolean;
  modified: boolean;
  returnDb: number;
  macro1Value: string;
  macro2Value: string;
}

export interface ProjectSessionFxSend {
  channelId: string;
  unitId: string;
  enabled: boolean;
  gainDb: number;
}

export interface ProjectSessionHarmony {
  channelId: string;
  contentRole: string;
  primaryHarmonyInstanceId: string;
  harmonyEnabled: boolean;
  key: string;
  scale: string;
  mode: string;
  voice1: string;
  voice2: string;
  harmonyLevelDb: number;
}

export interface ProjectSessionRecordedTake {
  id: string;
  path: string;
  tap: string;
  sampleRate: number;
  channelCount: number;
  frameCount: number;
  replayWithNeutralInserts: boolean;
  partial: boolean;
}

export function snapshotToProjectSession(snapshot: MixerSnapshot): ProjectSessionDocument {
  return {
    schemaVersion: 1,
    projectId: slugProjectId(snapshot.projectName),
    media: mediaRefs(snapshot),
    channels: snapshot.channels.map(channelToSession),
    fxUnits: snapshot.fxUnits.map((unit) => fxUnitToSession(unit, snapshot)),
    fxSends: snapshot.channels.flatMap(channelFxSends),
    channelHarmony: snapshot.channels.filter((channel) => channel.role === "vocal").map((channel) => ({
      channelId: channel.id,
      contentRole: channel.role,
      primaryHarmonyInstanceId: snapshot.harmony.primaryInstanceId,
      harmonyEnabled: Boolean(channel.harmonyEnabled),
      key: snapshot.harmony.key,
      scale: snapshot.harmony.scale,
      mode: snapshot.harmony.mode,
      voice1: snapshot.harmony.voice1,
      voice2: snapshot.harmony.voice2,
      harmonyLevelDb: snapshot.harmony.levelDb
    })),
    plugins: [],
    recordedTakes: snapshot.recording.takes.map(takeToSession)
  };
}

export function serializeProjectSession(snapshot: MixerSnapshot): string {
  return `${JSON.stringify(snapshotToProjectSession(snapshot), null, 2)}\n`;
}

export function projectSessionToSnapshot(content: string, baseSnapshot: MixerSnapshot): MixerSnapshot {
  const document = JSON.parse(content) as Partial<ProjectSessionDocument>;
  if (document.schemaVersion !== 1 || typeof document.projectId !== "string" || !Array.isArray(document.channels)) {
    throw new Error("Invalid Local Audio Mixer project document");
  }

  const snapshot = structuredClone(baseSnapshot);
  snapshot.projectName = document.projectId || snapshot.projectName;
  const savedChannels = document.channels ?? [];
  snapshot.channels = snapshot.channels.map((channel) => {
    const saved = savedChannels.find((item) => item.id === channel.id);
    return saved ? applyChannelSession(channel, saved) : channel;
  });
  const addedChannels = savedChannels
    .filter((saved) => !snapshot.channels.some((channel) => channel.id === saved.id))
    .map((saved) => channelSessionToSnapshot(saved, baseSnapshot));
  if (addedChannels.length > 0) {
    const masterIndex = snapshot.channels.findIndex((channel) => channel.kind === "master");
    const insertAt = masterIndex < 0 ? snapshot.channels.length : masterIndex;
    snapshot.channels = [
      ...snapshot.channels.slice(0, insertAt),
      ...addedChannels,
      ...snapshot.channels.slice(insertAt)
    ];
  }
  snapshot.fxUnits = snapshot.fxUnits.map((unit) => {
    const saved = document.fxUnits?.find((item) => item.unitId === unit.id);
    return saved ? applyFxUnitSession(unit, saved) : unit;
  });
  snapshot.channels = snapshot.channels.map((channel) => ({
    ...channel,
    sends: Object.fromEntries(Object.entries(channel.sends).map(([unitId, send]) => {
      const saved = document.fxSends?.find((item) => item.channelId === channel.id && item.unitId === unitId);
      return [unitId, saved ? { enabled: Boolean(saved.enabled), gainDb: Number(saved.gainDb) } : send];
    })) as ChannelState["sends"]
  }));
  const savedHarmony = document.channelHarmony?.find((item) => item.channelId === "voice") ?? document.channelHarmony?.[0];
  if (savedHarmony) {
    snapshot.harmony = {
      ...snapshot.harmony,
      enabled: Boolean(savedHarmony.harmonyEnabled),
      effectiveEnabled: Boolean(savedHarmony.harmonyEnabled),
      primaryInstanceId: savedHarmony.primaryHarmonyInstanceId,
      key: savedHarmony.key,
      scale: savedHarmony.scale,
      mode: savedHarmony.mode,
      voice1: savedHarmony.voice1,
      voice2: savedHarmony.voice2,
      levelDb: Number(savedHarmony.harmonyLevelDb),
      pending: false,
      error: ""
    };
  }
  snapshot.recording = {
    ...snapshot.recording,
    status: "idle",
    error: "",
    armedChannelIds: snapshot.channels.filter((channel) => channel.kind === "source" && channel.recordArm).map((channel) => channel.id),
    takes: (document.recordedTakes ?? []).flatMap(takeSessionToSnapshot),
    takeDirectory: firstTakeDirectory(document.recordedTakes ?? []) || snapshot.recording.takeDirectory
  };
  snapshot.selectedChannelId = snapshot.channels.some((channel) => channel.id === snapshot.selectedChannelId)
    ? snapshot.selectedChannelId
    : snapshot.channels[0]?.id ?? "";
  snapshot.channels = snapshot.channels.map((channel) => ({
    ...channel,
    selected: channel.id === snapshot.selectedChannelId,
    harmonyEnabled: channel.role === "vocal" ? snapshot.harmony.enabled : channel.harmonyEnabled
  }));
  snapshot.eqBands = structuredClone(snapshot.channels.find((channel) => channel.id === snapshot.selectedChannelId)?.eqBands ?? snapshot.eqBands);
  return snapshot;
}

function channelToSession(channel: ChannelState): ProjectSessionChannel {
  return {
    id: channel.id,
    name: channel.name,
    kind: channel.kind,
    role: channel.role,
    sourceUid: channel.kind === "source" ? channel.source : "",
    enabled: channel.enabled,
    muted: channel.mute,
    solo: channel.solo,
    monitor: Boolean(channel.monitor),
    recordArm: Boolean(channel.recordArm),
    eqEnabled: channel.processing.eq,
    noiseEnabled: channel.processing.noise,
    compEnabled: channel.processing.comp,
    deEsserEnabled: channel.processing.deEsser,
    insertFxEnabled: channel.processing.insertFx,
    gainDb: channel.trimDb,
    faderDb: channel.faderDb,
    pan: channel.pan / 100,
    noiseThresholdDb: channel.dynamics.noise.thresholdDb,
    noiseRangeDb: channel.dynamics.noise.rangeDb,
    noiseHoldMs: channel.dynamics.noise.holdMs,
    noiseReleaseMs: channel.dynamics.noise.releaseMs,
    compThresholdDb: channel.dynamics.compressor.thresholdDb,
    compRatio: channel.dynamics.compressor.ratio,
    compAttackMs: channel.dynamics.compressor.attackMs,
    compReleaseMs: channel.dynamics.compressor.releaseMs
  };
}

function applyChannelSession(channel: ChannelState, saved: ProjectSessionChannel): ChannelState {
  return {
    ...channel,
    name: saved.name || channel.name,
    source: saved.sourceUid || channel.source,
    enabled: Boolean(saved.enabled),
    mute: Boolean(saved.muted),
    solo: Boolean(saved.solo),
    monitor: Boolean(saved.monitor),
    recordArm: Boolean(saved.recordArm),
    trimDb: Number(saved.gainDb),
    faderDb: clamp(Number(saved.faderDb), -10, 10),
    pan: Number(saved.pan) * 100,
    processing: {
      ...channel.processing,
      eq: Boolean(saved.eqEnabled),
      noise: channel.role === "vocal" ? true : Boolean(saved.noiseEnabled),
      comp: Boolean(saved.compEnabled),
      deEsser: Boolean(saved.deEsserEnabled),
      insertFx: Boolean(saved.insertFxEnabled)
    },
    dynamics: {
      ...channel.dynamics,
      noise: {
        ...channel.dynamics.noise,
        thresholdDb: Number(saved.noiseThresholdDb),
        rangeDb: Number(saved.noiseRangeDb),
        holdMs: Number.isFinite(Number(saved.noiseHoldMs)) ? Number(saved.noiseHoldMs) : channel.dynamics.noise.holdMs,
        releaseMs: Number.isFinite(Number(saved.noiseReleaseMs)) ? Number(saved.noiseReleaseMs) : channel.dynamics.noise.releaseMs
      },
      compressor: {
        ...channel.dynamics.compressor,
        thresholdDb: Number(saved.compThresholdDb),
        ratio: Number(saved.compRatio),
        attackMs: Number.isFinite(Number(saved.compAttackMs)) ? Number(saved.compAttackMs) : channel.dynamics.compressor.attackMs,
        releaseMs: Number.isFinite(Number(saved.compReleaseMs)) ? Number(saved.compReleaseMs) : channel.dynamics.compressor.releaseMs
      }
    }
  };
}

function channelSessionToSnapshot(saved: ProjectSessionChannel, baseSnapshot: MixerSnapshot): ChannelState {
  const template = baseSnapshot.channels.find((channel) => channel.role === saved.role) ??
    baseSnapshot.channels.find((channel) => channel.kind === "source") ??
    baseSnapshot.channels[0];
  return applyChannelSession({
    ...(template ? structuredClone(template) : fallbackChannel()),
    id: saved.id,
    name: saved.name || saved.id,
    kind: parseKind(saved.kind),
    role: parseRole(saved.role),
    selected: false,
    meter: { left: -60, right: -60, clip: false }
  }, saved);
}

function fallbackChannel(): ChannelState {
  return {
    id: "channel",
    name: "CHANNEL",
    source: "",
    kind: "source",
    role: "music",
    selected: false,
    enabled: true,
    trimDb: 0,
    pan: 0,
    faderDb: 0,
    mute: false,
    solo: false,
    monitor: false,
    recordArm: false,
    processing: { eq: true, comp: false, noise: false, deEsser: false, insertFx: false },
    dynamics: {
      noise: { thresholdDb: -38, rangeDb: -50, holdMs: 30, releaseMs: 160 },
      compressor: { thresholdDb: -18, ratio: 3, attackMs: 10, releaseMs: 120 },
      deEsser: { frequencyHz: 6000, thresholdDb: -24, maxReductionDb: 6 }
    },
    sends: { "fx-a": { enabled: false, gainDb: -60 }, "fx-b": { enabled: false, gainDb: -60 } },
    eqBands: [],
    meter: { left: -60, right: -60, clip: false }
  };
}

function parseKind(kind: string): ChannelKind {
  return kind === "group" || kind === "master" ? kind : "source";
}

function parseRole(role: string): ChannelRole {
  if (role === "system" || role === "vocal" || role === "instrument" || role === "music" || role === "group" || role === "master") return role;
  return "music";
}

function applyFxUnitSession(unit: FxUnitState, saved: ProjectSessionFxUnit): FxUnitState {
  return {
    ...unit,
    programId: Number(saved.programId),
    revision: Number(saved.revision),
    enabled: Boolean(saved.enabled),
    modified: Boolean(saved.modified),
    returnDb: Number(saved.returnDb),
    pending: false,
    error: ""
  };
}

function fxUnitToSession(unit: FxUnitState, snapshot: MixerSnapshot): ProjectSessionFxUnit {
  const program = snapshot.programs.find((item) => item.id === unit.programId);
  return {
    unitId: unit.id,
    programId: unit.programId,
    processorType: program?.family ?? "",
    revision: unit.revision,
    enabled: unit.enabled,
    modified: unit.modified,
    returnDb: unit.returnDb,
    macro1Value: program?.macro1.value ?? "",
    macro2Value: program?.macro2.value ?? ""
  };
}

function channelFxSends(channel: ChannelState): ProjectSessionFxSend[] {
  return Object.entries(channel.sends).map(([unitId, send]) => ({
    channelId: channel.id,
    unitId,
    enabled: send.enabled,
    gainDb: send.gainDb
  }));
}

function takeToSession(take: RecordedTakeState): ProjectSessionRecordedTake {
  return {
    id: take.id,
    path: take.path,
    tap: take.tap,
    sampleRate: take.sampleRate,
    channelCount: take.channels,
    frameCount: take.frames,
    replayWithNeutralInserts: take.replayWithNeutralInserts,
    partial: take.partial
  };
}

function takeSessionToSnapshot(take: ProjectSessionRecordedTake): RecordedTakeState[] {
  if (!take || typeof take.id !== "string" || typeof take.path !== "string") return [];
  return [{
    id: take.id,
    path: take.path,
    tap: parseTap(take.tap),
    sampleRate: Number(take.sampleRate) || 48000,
    channels: Number(take.channelCount) || 2,
    frames: Number(take.frameCount) || 0,
    replayWithNeutralInserts: Boolean(take.replayWithNeutralInserts),
    partial: Boolean(take.partial)
  }];
}

function parseTap(tap: string): RecordingTap {
  if (tap === "dry" || tap === "processed" || tap === "master") return tap;
  return "master";
}

function firstTakeDirectory(takes: ProjectSessionRecordedTake[]) {
  const path = takes.find((take) => typeof take.path === "string" && take.path.includes("/"))?.path;
  return path ? path.slice(0, path.lastIndexOf("/")) : "";
}

function mediaRefs(snapshot: MixerSnapshot) {
  return snapshot.channels
    .filter((channel) => channel.role === "music" && channel.source.trim() !== "")
    .map((channel) => ({ id: channel.id, path: channel.source, missing: false }));
}

function slugProjectId(projectName: string) {
  const slug = projectName.trim().toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
  return slug || "local-audio-mixer";
}

function clamp(value: number, min: number, max: number) {
  if (!Number.isFinite(value)) return min;
  return Math.max(min, Math.min(max, value));
}

File: native/engine/src/engine/SessionDocument.cpp
#include "engine/SessionDocument.hpp"

#include <fstream>
#include <regex>
#include <sstream>
#include <system_error>

namespace localmixer::engine {
namespace {

std::string escapeJson(std::string_view value) {
  std::string escaped;
  for (const auto ch : value) {
    if (ch == '"' || ch == '\\') escaped += '\\';
    escaped += ch;
  }
  return escaped;
}

std::optional<std::string> readFile(const std::filesystem::path& path) {
  std::ifstream file(path);
  if (!file) return std::nullopt;
  std::ostringstream buffer;
  buffer << file.rdbuf();
  return buffer.str();
}

bool writeFile(const std::filesystem::path& path, std::string_view content) {
  std::ofstream file(path, std::ios::trunc);
  if (!file) return false;
  file << content;
  return static_cast<bool>(file);
}

std::optional<std::string> matchString(std::string_view json, const std::string& key) {
  const std::regex pattern("\"" + key + "\"\\s*:\\s*\"([^\"]*)\"");
  std::cmatch match;
  if (!std::regex_search(json.data(), json.data() + json.size(), match, pattern)) return std::nullopt;
  return match[1].str();
}

std::optional<std::uint32_t> matchU32(std::string_view json, const std::string& key) {
  const std::regex pattern("\"" + key + "\"\\s*:\\s*(\\d+)");
  std::cmatch match;
  if (!std::regex_search(json.data(), json.data() + json.size(), match, pattern)) return std::nullopt;
  return static_cast<std::uint32_t>(std::stoul(match[1].str()));
}

}  // namespace

std::string serializeSession(const SessionDocument& document) {
  std::ostringstream json;
  json << "{\n  \"schemaVersion\": " << document.schemaVersion << ",\n";
  json << "  \"projectId\": \"" << escapeJson(document.projectId) << "\",\n";
  json << "  \"media\": [";
  for (std::size_t index = 0; index < document.media.size(); index += 1) {
    const auto& media = document.media[index];
    json << (index == 0 ? "\n" : ",\n");
    json << "    {\"id\": \"" << escapeJson(media.id) << "\", \"path\": \""
         << escapeJson(media.path.string()) << "\", \"missing\": " << (media.missing ? "true" : "false") << "}";
  }
  if (!document.media.empty()) json << "\n  ";
  json << "],\n";
  json << "  \"channels\": [";
  for (std::size_t index = 0; index < document.channels.size(); index += 1) {
    const auto& channel = document.channels[index];
    json << (index == 0 ? "\n" : ",\n");
    json << "    {\"id\": \"" << escapeJson(channel.id)
         << "\", \"name\": \"" << escapeJson(channel.name)
         << "\", \"kind\": \"" << escapeJson(channel.kind)
         << "\", \"role\": \"" << escapeJson(channel.role)
         << "\", \"sourceUid\": \"" << escapeJson(channel.sourceUid)
         << "\", \"enabled\": " << (channel.enabled ? "true" : "false")
         << ", \"muted\": " << (channel.muted ? "true" : "false")
         << ", \"solo\": " << (channel.solo ? "true" : "false")
         << ", \"monitor\": " << (channel.monitor ? "true" : "false")
         << ", \"recordArm\": " << (channel.recordArm ? "true" : "false")
         << ", \"eqEnabled\": " << (channel.eqEnabled ? "true" : "false")
         << ", \"noiseEnabled\": " << (channel.noiseEnabled ? "true" : "false")
         << ", \"compEnabled\": " << (channel.compEnabled ? "true" : "false")
         << ", \"insertFxEnabled\": " << (channel.insertFxEnabled ? "true" : "false")
         << ", \"gainDb\": " << channel.gainDb
         << ", \"faderDb\": " << channel.faderDb
         << ", \"pan\": " << channel.pan
         << ", \"noiseThresholdDb\": " << channel.noiseThresholdDb
         << ", \"noiseRangeDb\": " << channel.noiseRangeDb
         << ", \"compThresholdDb\": " << channel.compThresholdDb
         << ", \"compRatio\": " << channel.compRatio << "}";
  }
  if (!document.channels.empty()) json << "\n  ";
  json << "],\n";
  json << "  \"fxUnits\": [";
  for (std::size_t index = 0; index < document.fxUnits.size(); index += 1) {
    const auto& unit = document.fxUnits[index];
    json << (index == 0 ? "\n" : ",\n");
    json << "    {\"unitId\": \"" << escapeJson(unit.unitId) << "\", \"programId\": " << unit.programId
         << ", \"processorType\": \"" << escapeJson(unit.processorType) << "\", \"revision\": " << unit.revision
         << ", \"enabled\": " << (unit.enabled ? "true" : "false")
         << ", \"modified\": " << (unit.modified ? "true" : "false")
         << ", \"returnDb\": " << unit.returnDb
         << ", \"macro1Value\": \"" << escapeJson(unit.macro1Value)
         << "\", \"macro2Value\": \"" << escapeJson(unit.macro2Value) << "\"}";
  }
  if (!document.fxUnits.empty()) json << "\n  ";
  json << "],\n";
  json << "  \"fxSends\": [";
  for (std::size_t index = 0; index < document.fxSends.size(); index += 1) {
    const auto& send = document.fxSends[index];
    json << (index == 0 ? "\n" : ",\n");
    json << "    {\"channelId\": \"" << escapeJson(send.channelId) << "\", \"unitId\": \"" << escapeJson(send.unitId)
         << "\", \"enabled\": " << (send.enabled ? "true" : "false")
         << ", \"gainDb\": " << send.gainDb << "}";
  }
  if (!document.fxSends.empty()) json << "\n  ";
  json << "],\n";
  json << "  \"channelHarmony\": [";
  for (std::size_t index = 0; index < document.channelHarmony.size(); index += 1) {
    const auto& harmony = document.channelHarmony[index];
    json << (index == 0 ? "\n" : ",\n");
    json << "    {\"channelId\": \"" << escapeJson(harmony.channelId) << "\", \"contentRole\": \"" << escapeJson(harmony.contentRole)
         << "\", \"primaryHarmonyInstanceId\": \"" << escapeJson(harmony.primaryHarmonyInstanceId)
         << "\", \"harmonyEnabled\": " << (harmony.harmonyEnabled ? "true" : "false")
         << ", \"key\": \"" << escapeJson(harmony.key)
         << "\", \"scale\": \"" << escapeJson(harmony.scale)
         << "\", \"mode\": \"" << escapeJson(harmony.mode)
         << "\", \"voice1\": \"" << escapeJson(harmony.voice1)
         << "\", \"voice2\": \"" << escapeJson(harmony.voice2)
         << "\", \"harmonyLevelDb\": " << harmony.harmonyLevelDb << "}";
  }
  if (!document.channelHarmony.empty()) json << "\n  ";
  json << "],\n";
  json << "  \"plugins\": [";
  for (std::size_t index = 0; index < document.plugins.size(); index += 1) {
    const auto& plugin = document.plugins[index];
    json << (index == 0 ? "\n" : ",\n");
    json << "    {\"instanceId\": \"" << escapeJson(plugin.instanceId)
         << "\", \"identifier\": \"" << escapeJson(plugin.identifier)
         << "\", \"version\": \"" << escapeJson(plugin.version)
         << "\", \"missing\": " << (plugin.missing ? "true" : "false")
         << ", \"stateBase64\": \"" << escapeJson(plugin.stateBase64) << "\"}";
  }
  if (!document.plugins.empty()) json << "\n  ";
  json << "],\n";
  json << "  \"recordedTakes\": [";
  for (std::size_t index = 0; index < document.recordedTakes.size(); index += 1) {
    const auto& take = document.recordedTakes[index];
    json << (index == 0 ? "\n" : ",\n");
    json << "    {\"id\": \"" << escapeJson(take.id)
         << "\", \"path\": \"" << escapeJson(take.path.string())
         << "\", \"tap\": \"" << escapeJson(take.tap)
         << "\", \"sampleRate\": " << take.sampleRate
         << ", \"channels\": " << take.channels
         << ", \"frames\": " << take.frames
         << ", \"replayWithNeutralInserts\": " << (take.replayWithNeutralInserts ? "true" : "false")
         << ", \"partial\": " << (take.partial ? "true" : "false") << "}";
  }
  if (!document.recordedTakes.empty()) json << "\n  ";
  json << "]\n}\n";
  return json.str();
}

SessionLoadResult parseSession(std::string_view json) {
  auto schemaVersion = matchU32(json, "schemaVersion");
  auto projectId = matchString(json, "projectId");
  if (!schemaVersion.has_value() || !projectId.has_value() || projectId->empty()) {
    return {.error = SessionError::invalidDocument};
  }
  if (*schemaVersion > kCurrentSessionSchemaVersion) return {.error = SessionError::newerSchema};

  SessionDocument document{.schemaVersion = *schemaVersion, .projectId = *projectId};
  const std::regex mediaPattern("\\{\"id\"\\s*:\\s*\"([^\"]*)\",\\s*\"path\"\\s*:\\s*\"([^\"]*)\",\\s*\"missing\"\\s*:\\s*(true|false)\\}");
  for (auto it = std::cregex_iterator(json.data(), json.data() + json.size(), mediaPattern);
       it != std::cregex_iterator();
       ++it) {
    document.media.push_back(SessionMediaRef{
      .id = (*it)[1].str(),
      .path = (*it)[2].str(),
      .missing = (*it)[3].str() == "true",
    });
  }

  const std::regex channelPattern("\\{\"id\"\\s*:\\s*\"([^\"]*)\",\\s*\"name\"\\s*:\\s*\"([^\"]*)\",\\s*\"kind\"\\s*:\\s*\"([^\"]*)\",\\s*\"role\"\\s*:\\s*\"([^\"]*)\",\\s*\"sourceUid\"\\s*:\\s*\"([^\"]*)\",\\s*\"enabled\"\\s*:\\s*(true|false),\\s*\"muted\"\\s*:\\s*(true|false),\\s*\"solo\"\\s*:\\s*(true|false),\\s*\"monitor\"\\s*:\\s*(true|false),\\s*\"recordArm\"\\s*:\\s*(true|false),\\s*\"eqEnabled\"\\s*:\\s*(true|false),\\s*\"noiseEnabled\"\\s*:\\s*(true|false),\\s*\"compEnabled\"\\s*:\\s*(true|false),\\s*\"insertFxEnabled\"\\s*:\\s*(true|false),\\s*\"gainDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?),\\s*\"faderDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?),\\s*\"pan\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?),\\s*\"noiseThresholdDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?),\\s*\"noiseRangeDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?),\\s*\"compThresholdDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?),\\s*\"compRatio\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?)\\}");
  for (auto it = std::cregex_iterator(json.data(), json.data() + json.size(), channelPattern);
       it != std::cregex_iterator();
       ++it) {
    document.channels.push_back(SessionChannelState{
      .id = (*it)[1].str(),
      .name = (*it)[2].str(),
      .kind = (*it)[3].str(),
      .role = (*it)[4].str(),
      .sourceUid = (*it)[5].str(),
      .enabled = (*it)[6].str() == "true",
      .muted = (*it)[7].str() == "true",
      .solo = (*it)[8].str() == "true",
      .monitor = (*it)[9].str() == "true",
      .recordArm = (*it)[10].str() == "true",
      .eqEnabled = (*it)[11].str() == "true",
      .noiseEnabled = (*it)[12].str() == "true",
      .compEnabled = (*it)[13].str() == "true",
      .insertFxEnabled = (*it)[14].str() == "true",
      .gainDb = std::stof((*it)[15].str()),
      .faderDb = std::stof((*it)[16].str()),
      .pan = std::stof((*it)[17].str()),
      .noiseThresholdDb = std::stof((*it)[18].str()),
      .noiseRangeDb = std::stof((*it)[19].str()),
      .compThresholdDb = std::stof((*it)[20].str()),
      .compRatio = std::stof((*it)[21].str()),
    });
  }

  const std::regex fxUnitPattern("\\{\"unitId\"\\s*:\\s*\"([^\"]*)\",\\s*\"programId\"\\s*:\\s*(\\d+),\\s*\"processorType\"\\s*:\\s*\"([^\"]*)\",\\s*\"revision\"\\s*:\\s*(\\d+),\\s*\"enabled\"\\s*:\\s*(true|false),\\s*\"modified\"\\s*:\\s*(true|false),\\s*\"returnDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?),\\s*\"macro1Value\"\\s*:\\s*\"([^\"]*)\",\\s*\"macro2Value\"\\s*:\\s*\"([^\"]*)\"\\}");
  for (auto it = std::cregex_iterator(json.data(), json.data() + json.size(), fxUnitPattern);
       it != std::cregex_iterator();
       ++it) {
    document.fxUnits.push_back(SessionFxUnitState{
      .unitId = (*it)[1].str(),
      .programId = static_cast<std::uint32_t>(std::stoul((*it)[2].str())),
      .processorType = (*it)[3].str(),
      .revision = static_cast<std::uint64_t>(std::stoull((*it)[4].str())),
      .enabled = (*it)[5].str() == "true",
      .modified = (*it)[6].str() == "true",
      .returnDb = std::stof((*it)[7].str()),
      .macro1Value = (*it)[8].str(),
      .macro2Value = (*it)[9].str(),
    });
  }

  const std::regex sendPattern("\\{\"channelId\"\\s*:\\s*\"([^\"]*)\",\\s*\"unitId\"\\s*:\\s*\"([^\"]*)\",\\s*\"enabled\"\\s*:\\s*(true|false),\\s*\"gainDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?)\\}");
  for (auto it = std::cregex_iterator(json.data(), json.data() + json.size(), sendPattern);
       it != std::cregex_iterator();
       ++it) {
    document.fxSends.push_back(SessionFxSendAssignment{
      .channelId = (*it)[1].str(),
      .unitId = (*it)[2].str(),
      .enabled = (*it)[3].str() == "true",
      .gainDb = std::stof((*it)[4].str()),
    });
  }

  const std::regex harmonyPattern("\\{\"channelId\"\\s*:\\s*\"([^\"]*)\",\\s*\"contentRole\"\\s*:\\s*\"([^\"]*)\",\\s*\"primaryHarmonyInstanceId\"\\s*:\\s*\"([^\"]*)\",\\s*\"harmonyEnabled\"\\s*:\\s*(true|false),\\s*\"key\"\\s*:\\s*\"([^\"]*)\",\\s*\"scale\"\\s*:\\s*\"([^\"]*)\",\\s*\"mode\"\\s*:\\s*\"([^\"]*)\",\\s*\"voice1\"\\s*:\\s*\"([^\"]*)\",\\s*\"voice2\"\\s*:\\s*\"([^\"]*)\",\\s*\"harmonyLevelDb\"\\s*:\\s*(-?\\d+(?:\\.\\d+)?)\\}");
  for (auto it = std::cregex_iterator(json.data(), json.data() + json.size(), harmonyPattern);
       it != std::cregex_iterator();
       ++it) {
    document.channelHarmony.push_back(SessionChannelHarmonyState{
      .channelId = (*it)[1].str(),
      .contentRole = (*it)[2].str(),
      .primaryHarmonyInstanceId = (*it)[3].str(),
      .harmonyEnabled = (*it)[4].str() == "true",
      .key = (*it)[5].str(),
      .scale = (*it)[6].str(),
      .mode = (*it)[7].str(),
      .voice1 = (*it)[8].str(),
      .voice2 = (*it)[9].str(),
      .harmonyLevelDb = std::stof((*it)[10].str()),
    });
  }

  const std::regex pluginPattern("\\{\"instanceId\"\\s*:\\s*\"([^\"]*)\",\\s*\"identifier\"\\s*:\\s*\"([^\"]*)\",\\s*\"version\"\\s*:\\s*\"([^\"]*)\",\\s*\"missing\"\\s*:\\s*(true|false),\\s*\"stateBase64\"\\s*:\\s*\"([^\"]*)\"\\}");
  for (auto it = std::cregex_iterator(json.data(), json.data() + json.size(), pluginPattern);
       it != std::cregex_iterator();
       ++it) {
    document.plugins.push_back(SessionPluginState{
      .instanceId = (*it)[1].str(),
      .identifier = (*it)[2].str(),
      .version = (*it)[3].str(),
      .missing = (*it)[4].str() == "true",
      .stateBase64 = (*it)[5].str(),
    });
  }

  const std::regex takePattern("\\{\"id\"\\s*:\\s*\"([^\"]*)\",\\s*\"path\"\\s*:\\s*\"([^\"]*)\",\\s*\"tap\"\\s*:\\s*\"([^\"]*)\",\\s*\"sampleRate\"\\s*:\\s*(\\d+),\\s*\"channels\"\\s*:\\s*(\\d+),\\s*\"frames\"\\s*:\\s*(\\d+),\\s*\"replayWithNeutralInserts\"\\s*:\\s*(true|false),\\s*\"partial\"\\s*:\\s*(true|false)\\}");
  for (auto it = std::cregex_iterator(json.data(), json.data() + json.size(), takePattern);
       it != std::cregex_iterator();
       ++it) {
    document.recordedTakes.push_back(SessionRecordedTake{
      .id = (*it)[1].str(),
      .path = (*it)[2].str(),
      .tap = (*it)[3].str(),
      .sampleRate = static_cast<std::uint32_t>(std::stoul((*it)[4].str())),
      .channels = static_cast<std::uint16_t>(std::stoul((*it)[5].str())),
      .frames = static_cast<std::uint64_t>(std::stoull((*it)[6].str())),
      .replayWithNeutralInserts = (*it)[7].str() == "true",
      .partial = (*it)[8].str() == "true",
    });
  }
  return {.document = document};
}

bool saveSessionAtomic(const std::filesystem::path& path, const SessionDocument& document) {
  if (document.schemaVersion != kCurrentSessionSchemaVersion || document.projectId.empty()) return false;
  std::error_code error;
  std::filesystem::create_directories(path.parent_path(), error);
  if (error) return false;

  const auto tempPath = path.string() + ".tmp";
  const auto backupPath = path.string() + ".bak";
  if (!writeFile(tempPath, serializeSession(document))) return false;
  if (std::filesystem::exists(path, error)) {
    std::filesystem::copy_file(path, backupPath, std::filesystem::copy_options::overwrite_existing, error);
    if (error) return false;
  }
  std::filesystem::rename(tempPath, path, error);
  return !error;
}

SessionLoadResult loadSession(const std::filesystem::path& path) {
  const auto content = readFile(path);
  if (!content.has_value()) return {.error = SessionError::ioError};
  return parseSession(*content);
}

std::filesystem::path autosavePathFor(const std::filesystem::path& sessionPath) {
  auto autosavePath = sessionPath;
  autosavePath += ".autosave";
  return autosavePath;
}

std::optional<std::filesystem::path> recoveryCandidateFor(const std::filesystem::path& sessionPath) {
  const auto autosavePath = autosavePathFor(sessionPath);
  std::error_code error;
  if (std::filesystem::exists(autosavePath, error) && std::filesystem::is_regular_file(autosavePath, error)) {
    return autosavePath;
  }
  return std::nullopt;
}

}  // namespace localmixer::engine
